okio-files/src/commonMain/kotlin/okio/Filesystem.kt

```diff
@@ -109,6 +109,12 @@ abstract class Filesystem {
   @Throws(IOException::class)
   abstract fun delete(path: Path)
 
+  /**
+   * Returns a writable temporary directory on the current file system.
+   * This is the java.io.tmpdir system property on the JVM platform and the TMPDIR environment variable on the POSIX platform
+   */
+  abstract fun temporaryDirectory(): Path
```

Comment:
This doesn't feel like it belongs here since it's not a property intrinsic to a file system. What's the temporary directory for a read-only zip file mounted as a filesystem? A temp dir is a property intrinsic to an operating system or platform.

Reply:
Seems like it's only used by tests? Maybe just an actual/expect in there to start with?

Reply:
I wanna use this to write multiplatform tests that use files. For example, tests of DiskLruCache.

A zip file is not an exemplar file system. For example, it doesn't support random access writing.

Reply:
(and if you are doing a Zip filesystem, you can just return Filesystem.SYSTEM.temporaryDirectory().

Reply:
That argument doesn't sway me. I think you've only radicalized me further. Temporary directories aren't an intrinsic property of a file system, they're a property of an operating system. At best it's a path resolvable by `FileSystem.SYSTEM` but it is not intrinsic to all file systems. It belongs elsewhere.

Reply:
They are certainly confined to a file system, however. It would be an error to take the temporary directory of my operating system and use it to write to a `.zip` file.

I think what you really want is this:
```
  fun temporaryDirectory(): Path?
```

That way it’s optional, and we can document that `FileSystem.SYSTEM` has one.

Reply:
What about a Samba or NFS filesystem mounted over the network? Or exposing S3 as a filesystem either through HTTP calls or the AWS S3 SDK? A zip file just an easy example to litmus test the API against that should quickly allow us to identify what is something primitive to a filesystem and what is not.

In your DiskLruCache example the test should have no behavioral difference whether given the system FS or an in-memory FS. The test doesn't get to ask for a temporary directory, it's given an FS and a base Path. If the test runner (or parameterized factory or whatever) wants to use the system FS and the OS temporary directory it can, but for an in-memory FS the temporary path can be the root dir.

Reply:
@martinbonnin wanna pull this back from the public API? @JakeWharton and I will figure out the API in a follow-up.

Reply:
Sure, let me try

Reply:
I just made `temporaryDirectory` internal. I think that's enough to remove it from the public API. Let me know if you want me to bury it a bit deeper

Reply:
👍